File: VB/Example/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System.Web.Mvc
Imports Example.Models
Imports DevExpress.Web.Mvc
Imports DevExpress.Data.Filtering

Namespace Example.Controllers
	Public Class HomeController
		Inherits Controller
		Public Function Index() As ActionResult
			Return View()
		End Function
		Public Function MasterGridViewPartial() As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("masterGrid")
			If viewModel Is Nothing Then
				viewModel = CreateMasterGridViewModel()
			End If
			Return MasterGridActionCore(viewModel)
		End Function
		Public Function MasterGridViewSortingAction(ByVal column As GridViewColumnState, ByVal reset As Boolean) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("masterGrid")
			viewModel.SortBy(column, reset)
			Return MasterGridActionCore(viewModel)
		End Function
		Public Function MasterGridViewPagingAction(ByVal pager As GridViewPagerState) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("masterGrid")
			viewModel.Pager.Assign(pager)
			Return MasterGridActionCore(viewModel)
		End Function
		Public Function MasterGridActionCore(ByVal gridViewModel As GridViewModel) As ActionResult
            gridViewModel.ProcessCustomBinding(AddressOf MasterCustomBindingHandlers.GetDataRowCount, AddressOf MasterCustomBindingHandlers.GetData)
			Return PartialView("MasterGridViewPartial", gridViewModel)
		End Function

		Public Function DetailGridViewPartial(ByVal customerID As String) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("detailGrid" & customerID)
			If viewModel Is Nothing Then
				viewModel = CreateDetailGridViewModel(customerID)
			End If
			Return DetailGridActionCore(viewModel, customerID)
		End Function
		Public Function DetailGridViewPagingAction(ByVal pager As GridViewPagerState, ByVal customerID As String) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("detailGrid" & customerID)
			viewModel.Pager.Assign(pager)
			Return DetailGridActionCore(viewModel, customerID)
		End Function
		Public Function DetailGridViewSortingAction(ByVal column As GridViewColumnState, ByVal reset As Boolean, ByVal customerID As String) As ActionResult
			Dim viewModel = GridViewExtension.GetViewModel("detailGrid" & customerID)
			viewModel.SortBy(column, reset)
			Return DetailGridActionCore(viewModel, customerID)
		End Function
		Public Function DetailGridActionCore(ByVal gridViewModel As GridViewModel, ByVal customerID As String) As ActionResult
            gridViewModel.ProcessCustomBinding(AddressOf DetailCustomBindingHandlers.GetDataRowCount, AddressOf DetailCustomBindingHandlers.GetData)
			ViewData("CustomerID") = customerID
			Return PartialView("DetailGridViewPartial", gridViewModel)
		End Function

		Private Shared Function CreateMasterGridViewModel() As GridViewModel
			Dim viewModel = New GridViewModel()
			viewModel.KeyFieldName = "CustomerID"
			viewModel.Columns.Add("ContactName")
			viewModel.Columns.Add("CompanyName")
			viewModel.Columns.Add("City")
			viewModel.Columns.Add("Country")
			Return viewModel
		End Function
		Private Shared Function CreateDetailGridViewModel(ByVal customerID As String) As GridViewModel
			Dim viewModel = New GridViewModel()
			viewModel.KeyFieldName = "OrderID"
			viewModel.Columns.Add("OrderID")
			viewModel.Columns.Add("OrderDate")
			viewModel.Columns.Add("ShipName")
			viewModel.Columns.Add("ProductName")
			viewModel.Columns.Add("Quantity")
			viewModel.Columns.Add("UnitPrice")
			viewModel.FilterExpression = (New BinaryOperator("CustomerID", customerID)).ToString()
			Return viewModel
		End Function
	End Class
End Namespace

File: CS/Example/Controllers/HomeController.cs
using System.Web.Mvc;
using Example.Models;
using DevExpress.Web.Mvc;
using DevExpress.Data.Filtering;

namespace Example.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View();	
        }
		public ActionResult MasterGridViewPartial() {
            var viewModel = GridViewExtension.GetViewModel("masterGrid");
            if(viewModel == null)
                viewModel = CreateMasterGridViewModel();
            return MasterGridActionCore(viewModel);
        }
        public ActionResult MasterGridViewSortingAction(GridViewColumnState column, bool reset) {
            var viewModel = GridViewExtension.GetViewModel("masterGrid");
            viewModel.SortBy(column, reset);
            return MasterGridActionCore(viewModel);
        }
        public ActionResult MasterGridViewPagingAction(GridViewPagerState pager) {
            var viewModel = GridViewExtension.GetViewModel("masterGrid");
            viewModel.Pager.Assign(pager);
            return MasterGridActionCore(viewModel);
        }
        public ActionResult MasterGridActionCore(GridViewModel gridViewModel) {
            gridViewModel.ProcessCustomBinding(
                MasterCustomBindingHandlers.GetDataRowCount,
                MasterCustomBindingHandlers.GetData
            );
            return PartialView("MasterGridViewPartial", gridViewModel);
        }

        public ActionResult DetailGridViewPartial(string customerID) {
            var viewModel = GridViewExtension.GetViewModel("detailGrid" + customerID);
            if(viewModel == null)
                viewModel = CreateDetailGridViewModel(customerID);
            return DetailGridActionCore(viewModel, customerID);
        }
        public ActionResult DetailGridViewPagingAction(GridViewPagerState pager, string customerID) {
            var viewModel = GridViewExtension.GetViewModel("detailGrid" + customerID);
            viewModel.Pager.Assign(pager);
            return DetailGridActionCore(viewModel, customerID);
        }
        public ActionResult DetailGridViewSortingAction(GridViewColumnState column, bool reset, string customerID) {
            var viewModel = GridViewExtension.GetViewModel("detailGrid" + customerID);
            viewModel.SortBy(column, reset);
            return DetailGridActionCore(viewModel, customerID);
        }
        public ActionResult DetailGridActionCore(GridViewModel gridViewModel, string customerID) {
            gridViewModel.ProcessCustomBinding(
                DetailCustomBindingHandlers.GetDataRowCount,
                DetailCustomBindingHandlers.GetData
            );
            ViewData["CustomerID"] = customerID;
            return PartialView("DetailGridViewPartial", gridViewModel);
        }

        static GridViewModel CreateMasterGridViewModel() {
            var viewModel = new GridViewModel();
            viewModel.KeyFieldName = "CustomerID";
            viewModel.Columns.Add("ContactName");
            viewModel.Columns.Add("CompanyName");
            viewModel.Columns.Add("City");
            viewModel.Columns.Add("Country");
            return viewModel;
        }
        static GridViewModel CreateDetailGridViewModel(string customerID) {
            var viewModel = new GridViewModel();
            viewModel.KeyFieldName = "OrderID";
            viewModel.Columns.Add("OrderID");
            viewModel.Columns.Add("OrderDate");
            viewModel.Columns.Add("ShipName");
            viewModel.Columns.Add("ProductName");
            viewModel.Columns.Add("Quantity");
            viewModel.Columns.Add("UnitPrice");
            viewModel.FilterExpression = (new BinaryOperator("CustomerID", customerID)).ToString();
            return viewModel;
        }
	}
}